public/app/features/dashboard-scene/addToDashboard/addToDashboard.ts

```diff
@@ -0,0 +1,88 @@
+import { locationUtil, TimeRange } from '@grafana/data';
+import { config, locationService } from '@grafana/runtime';
+import { Panel } from '@grafana/schema';
+import { DASHBOARD_SCHEMA_VERSION } from 'app/features/dashboard/state/DashboardMigrator';
+import {
+  removeDashboardToFetchFromLocalStorage,
+  setDashboardToFetchFromLocalStorage,
+} from 'app/features/dashboard/state/initDashboard';
+import { DashboardDTO } from 'app/types';
+
+export enum GenericError {
+  UNKNOWN = 'unknown-error',
+  NAVIGATION = 'navigation-error',
+}
+
+export interface SubmissionError {
+  error: AddToDashboardError | GenericError;
+  message: string;
+}
+
+export enum AddToDashboardError {
+  FETCH_DASHBOARD = 'fetch-dashboard',
+  SET_DASHBOARD_LS = 'set-dashboard-ls-error',
+}
+
+interface AddPanelToDashboardOptions {
+  panel: Panel;
+  dashboardUid?: string;
+  openInNewTab?: boolean;
+  timeRange?: TimeRange;
+}
+
+export function addToDashboard({
+  panel,
+  dashboardUid,
+  openInNewTab,
+  timeRange,
+}: AddPanelToDashboardOptions): SubmissionError | undefined {
+  let dto: DashboardDTO = {
+    meta: {},
+    dashboard: {
+      title: '',
+      uid: '',
+      panels: [panel],
+      schemaVersion: DASHBOARD_SCHEMA_VERSION,
+    },
+  };
+
+  if (timeRange) {
+    const raw = timeRange.raw;
+    dto.dashboard.time = {
+      from: typeof raw.from === 'string' ? raw.from : raw.from.toISOString(),
+      to: typeof raw.to === 'string' ? raw.to : raw.to.toISOString(),
+    };
```

Comment:
During testing I've noticed that two history entries are created.
1. Add panel from Explore w/ time range last 1 h
2. Dashboard has time range "taken from explore" even thou, it's default is last 6h
3. Navigate back - time range changes on the dashboard from last 1h to last 6h
4. Navigate back - Discard is shown and one can navigate back to Explore.


It looks like a regression, compared with play.

Reply:
> 2. Dashboard has time range "taken from explore" even thou, it's default is last 6h

@dprokop this is how it works for me in play. The dashboard get's the same time range as was used in explore 

3. Sounds ok, you get the original dashboard time range. 

4. Sounds good? 

I guess the only change is you can undo the time range change by going back? that is true of all time range changes

Reply:
The problem is pretty much 3) - the rest is just flow description.